buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -114,6 +116,27 @@ class BuildPlugin implements Plugin<Project> {
                 throw new GradleException("Java ${minimumJava} or above is required to build Elasticsearch")
             }
 
+            // this block of code detecting buggy JDK 8 compiler versions can be removed when minimum Java version is incremented
+            assert minimumJava == JavaVersion.VERSION_1_8 : "Remove JDK compiler bug detection only applicable to JDK 8"
+            if (javaVersionEnum == JavaVersion.VERSION_1_8) {
+                if (Objects.equals("Oracle Corporation", javaVendor)) {
+                    def matcher = javaVersion =~ /1\.8\.0(?:_(\d+))?/
+                    if (matcher.matches()) {
+                        int update;
+                        if (matcher.group(1) == null) {
+                            update = 0
+                        } else {
+                            update = matcher.group(1).toInteger()
+                        }
+                        if (update < 40) {
+                            throw new GradleException("JDK ${javaVendor} ${javaVersion} has compiler bug JDK-8052388, update your JDK to at least 8u40")
+                        }
+                    } else {
+                        println "Unable to detect JDK update version for JDK ${javaVendor} ${javaVersion}; your JDK might be subject to compiler bug JDK-8052388!"
```

Comment:
I'm not sure if it's a good idea to print a warning in this case as it might lead to more confusion. If a user runs into JDK-8052388 when compiling and reports it, we can always refine the check above.